streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStoreTest.java

```diff
@@ -0,0 +1,312 @@
+        doShouldPassChangelogTopicNameToStateStoreSerde(defaultChangelogTopicName);
+    }
+
+    @SuppressWarnings("unchecked")
+    private void doShouldPassChangelogTopicNameToStateStoreSerde(final String changelogTopicName) {
+        // recreate store with mock serdes
+        final Serde<String> keySerde = mock(Serde.class);
+        final Serializer<String> keySerializer = mock(Serializer.class);
+        final Serde<String> valueSerde = mock(Serde.class);
+        final Serializer<String> valueSerializer = mock(Serializer.class);
+        final Deserializer<String> valueDeserializer = mock(Deserializer.class);
+        when(keySerde.serializer()).thenReturn(keySerializer);
+        when(valueSerde.serializer()).thenReturn(valueSerializer);
+        when(valueSerde.deserializer()).thenReturn(valueDeserializer);
+
+        store.close();
+        store = new MeteredVersionedKeyValueStore<>(
+            inner,
+            METRICS_SCOPE,
+            mockTime,
+            keySerde,
+            new NullableValueAndTimestampSerde<>(valueSerde)
+        );
+        store.init((StateStoreContext) context, store);
+
+        store.put(KEY, VALUE, TIMESTAMP);
+
+        verify(keySerializer).serialize(changelogTopicName, KEY);
+        verify(valueSerializer).serialize(changelogTopicName, VALUE);
+    }
+
+    @Test
+    public void shouldRecordMetricsOnInit() {
+        // init is called in setUp(). it suffices to verify one restore metric since all restore
+        // metrics are recorded by the same sensor, and the sensor is tested elsewhere.
+        assertThat((Double) getMetric("restore-rate").metricValue(), greaterThan(0.0));
+    }
+
+    @Test
+    public void shouldDelegateAndRecordMetricsOnPut() {
+        store.put(KEY, VALUE, TIMESTAMP);
+
+        verify(inner).put(RAW_KEY, RAW_VALUE_AND_TIMESTAMP);
+        assertThat((Double) getMetric("put-rate").metricValue(), greaterThan(0.0));
+    }
+
+    @Test
+    public void shouldDelegateAndRecordMetricsOnDelete() {
+        store.delete(KEY, TIMESTAMP);
+
+        verify(inner).delete(RAW_KEY, TIMESTAMP);
+        assertThat((Double) getMetric("delete-rate").metricValue(), greaterThan(0.0));
+    }
+
+    @Test
+    public void shouldDelegateAndRecordMetricsOnGet() {
+        store.get(KEY);
+
+        verify(inner).get(RAW_KEY);
+        assertThat((Double) getMetric("get-rate").metricValue(), greaterThan(0.0));
+    }
+
+    @Test
+    public void shouldDelegateAndRecordMetricsOnGetWithTimestamp() {
+        store.get(KEY, TIMESTAMP);
+
+        verify(inner).get(RAW_KEY, TIMESTAMP);
+        assertThat((Double) getMetric("get-rate").metricValue(), greaterThan(0.0));
+    }
+
+    @Test
+    public void shouldDelegateAndRecordMetricsOnFlush() {
+        store.flush();
+
+        verify(inner).flush();
+        assertThat((Double) getMetric("flush-rate").metricValue(), greaterThan(0.0));
+    }
+
+    @Test
+    public void shouldDelegateAndRemoveMetricsOnClose() {
```

Comment:
Do we need this test? Seem it's rather testing the wrapped-store?

Reply:
Not sure I follow. This test verifies that metrics are properly cleaned up when the store is closed, and this functionality is unique to this metered store layer. Maybe you mean that this test tests functionality which is inherited from `MeteredKeyValueStore`? That is true, but I think it is important to re-test here because we want to ensure that `MeteredVersionedKeyValueStore` properly cleans up its metrics as well, i.e., `MeteredVersionedKeyValueStore` is separate from `MeteredKeyValueStore`.

Reply:
Fair enough.